Add neureka support

lukamac opened the pull request.

lukamac committed:
Add neureka support similar to ne16

lukamac committed:
add neureka support to test app

lukamac committed:
Add weight mem source flag

lukamac committed:
Fix formatting

lukamac committed:
Fix strides and counters, remove stride2x2 and flag16 mode

lukamac committed:
Fix uninitialized L1 data

lukamac committed:
Fixup nnx_quant_t nnx_norm_t

lukamac committed:
Add Neureka weight roll/unroll script

lukamac committed:
Fix Ne16 weight rolling was unpacking to bits instead of 8

lukamac committed:
Fix generated arrays initialization

lukamac committed:
Fix Neureka weight unroll for 1x1 mode

lukamac committed:
Fix Arpan's name in the contributors

lukamac committed:
Remove WIEGHT_D0_STRIDE_MODE_1x1

lukamac committed:
Add multi-accelerator support and neureka as a target

lukamac committed:
Remove stride2x2 for neureka

lukamac committed:
Fix formatting

lukamac committed:
Skip invalid tests

lukamac committed:
Change invalid tes tskip to explicit pytest skip with reason

lukamac committed:
Add neureka test to the CI

lukamac committed:
Fix formatting

lukamac committed:
Add --print-tensors flag

lukamac committed:
Remove memcpys since it was a linker script bug

lukamac committed:
Add Application section to test's readme

lukamac committed:
Fix formatting

lukamac committed:
Fix accelerator name printing

lukamac committed:
Add input_signed to neureka

lukamac committed:
Add readme per accelerator

lukamac committed:
Remove ne16 input dim 2 stride calculation

lukamac committed:
Move common validators to NnxTest

lukamac committed:
Fix xored relu with unsigned instead of signed

lukamac committed:
Change gvsoc functions have a body only when running in gvsoc

lukamac committed:
Change remove %s from main strings

lukamac committed:
Fix output checking should be done before cluster close since data is in L1

lukamac committed:
Remove L2 copies of tensors

lukamac committed:
Replace pointer arithmetic with array indexing

lukamac committed:
Fix formatting

lukamac committed:
Fix Neureka name in readme

lukamac committed:
Pad cin

lukamac committed:
Align names with Neureka.py

Scheremo commented on ne16/README.md:
```diff
@@ -0,0 +1,37 @@
+    - [ ] Per-channel shift
+    - [x] Per-layer shift
+    - [ ] Rounding
+- [ ] Input type
+    - [x] uint8
+    - [ ] uint16
```
Shouldn't there also be int8?

lukamac replied:
for NE16, no

Scheremo commented on ne16/gvsoc/ne16_gvsoc.h:
```diff
@@ -42,13 +42,17 @@ typedef enum ne16_gvsoc_log_level_e {
 static void ne16_gvsoc_log_activate(ne16_dev_t *dev,
                                     ne16_gvsoc_log_level_e log_level,
                                     ne16_gvsoc_log_format_e format) {
+#if __PLATFORM__ == ARCHI_PLATFORM_GVSOC
```
This depends on PULP SDK, right? It might be good to add comment in the README for people using this library without PULP SDK.

lukamac replied:
Removed them from here and added them in the tests since the tests are done with pulp-sdk and the library user might not. Commit https://github.com/pulp-platform/pulp-nnx/pull/2/commits/9145445c16ce982fd1215f305389ee0ad68a0fe0

Scheremo commented on neureka/bsp/neureka_siracusa_bsp.c:
```diff
@@ -0,0 +1,95 @@
+#define NEUREKA_SIRACUSA_WEIGHT_MEM_MRAM_OFFSET (0x00000000)
+#define NEUREKA_SIRACUSA_WEIGHT_MEM_SRAM_OFFSET (0x00400000)
+
+void neureka_siracusa_cg_enable() {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR |=
+      NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_CG_EN;
+}
+
+void neureka_siracusa_cg_disable() {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR &=
+      ~NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_CG_EN;
+}
```
There are no clock gates for N-EUREKA in Siracusa :)

FrancescoConti replied:
👍  but we should still keep the stubs as that is definitely a bug, not a feature :D 
BTW isn't there the magic trick? Did you ever manage to make it work?

Scheremo replied:
About the stub: Yes! But I would take the opportunity to clarify that this won't make a functional difference.
The `magic trick` we discussed ~1 year ago? That one didn't really pull the power consumption down to the level I'd expect, and I didn't manage to reliably clear the resulting state :/ But I didn't go through RTL sim, so take it with a grain of salt.

lukamac replied:
Thanks for pointing this out! I prefer removing them if they are not working.

FrancescoConti replied:
I think it's much better to keep an empty stub than an #ifdef SIRACUSA_FIXED or something, but your call...

lukamac replied:
My idea is to have a different `neureka_<board>_bsp.c` file for each platform. If a future platform uses neureka and provides clock gating for it, it will be implemented.

Scheremo commented on neureka/hal/neureka.c:
```diff
@@ -18,15 +18,22 @@
 
-#endif // __NE16_ERROR_CODES_H__
\ No newline at end of file
+inline int neureka_task_queue_size(neureka_dev_t *dev) { return 2; }
```
:)

lukamac replied:
Yes?

Scheremo replied:
Is the task queue size something that derives from the neureka_dev_t struct? Is it something that might vary with future implementations? If not I think a `#define` might be a good solution worth considering.

FrancescoConti replied:
I believe this is the reflection of a hardware parameter that can be 2, 2, or sometimes 2. Just kidding. It can also be 1 (to define at accelerator design time) or, in theory, up to 4 with small changes, but in practice 2 works better than 1 at little added cost and it is very hard to find a use case for >2.
So I'd go for the `#define` :D

lukamac replied:
Changed here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/1a4f873327ea3b7271e0a313a34ce3661bad5d21

Scheremo commented on neureka/hal/neureka_task.c:
```diff
@@ -0,0 +1,201 @@
+#include "neureka_task_defs.h"
+#include "pulp_nnx_util.h"
+
+inline uint32_t neureka_get_tile_padding(uint32_t padding, uint32_t i_height,
```
NIT: Are you sure about force inling this? Looks like it's used in numerous places and it's a decent amount of code...

lukamac replied:
I will remove and let the compiler decide.

Scheremo commented on neureka/hal/neureka_task.c:
```diff
@@ -0,0 +1,201 @@
+                        : depthwise == 1  ? NEUREKA_FLAG_MODE_3x3_DW
+                                          : NEUREKA_FLAG_MODE_3x3;
+
+  task->data.cfg.conf0 |=
+      flag_input_signed << NEUREKA_SHIFT_FLAG_INPUT_SIGNED |
+      NEUREKA_FLAG_NORM_QUANT | quant.function | quant.mode |
+      (quant.shift_amount << 16) |
+      quant.flag_rounding << NEUREKA_SHIFT_ROUNDING | norm.mode |
+      norm.flag_bias << NEUREKA_SHIFT_FLAG_NORM_BIAS |
+      norm.flag_shift << NEUREKA_SHIFT_FLAG_NORM_SHIFT | NEUREKA_FLAG_USE_TCDM |
+      weights_offset_mode | flag_mode | (weights_bits - 1);
```
I'd put a link to the register map of N-EUREKA here. While this looks correct, it seems hard to verify without this information.

Scheremo commented on neureka/hal/neureka_task.c:
```diff
@@ -0,0 +1,201 @@
+  task->data.cfg.input_stride = input_stride;
+
+  const neureka_stride_t output_stride = {
+      .d0 = 32, // TODO: should depend on outbytes. Probably 32 / outbytes
```
I'd verify this TODO before merging

lukamac replied:
Todo verified. It is in fact not probably `32 / outbytes` but just 32, which is the output bandwidth in bytes. Added a define.

Scheremo commented on neureka/hal/neureka_task.c:
```diff
@@ -0,0 +1,201 @@
+  const uint16_t rem_Hi =
+      rem_Ho == 0 ? 0
+                  : (task->kernel_shape == 1 ? rem_Ho : rem_Ho + 2) -
+                        padding_bottom; // TODO: Check padding bottom
```
SEMI-NIT: This is a bit hard to follow, it might be worth it to extract the second ternary operator to help readability.

lukamac replied:
Removed the `rem_* == 0` here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/d9c45efe27e0918e620221a9b0839f8e5eb81d89
This is never true due to the actual implementation of the remainder function, which either gives remainder or divisor if the remainder is 0.

lukamac replied:
Renamed remainder here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/40657fd10da6f0998425e8917725f1a3a12f34a1

Scheremo commented on test/TestClasses.py:
```diff
@@ -32,6 +32,10 @@ def implies(a: bool, b: bool):
     return (not a) or b
 
 
+def xor(a: bool, b: bool):
+    return (a and not b) or (not a and b)
```
Why not ^?

lukamac replied:
For some reason I thought python didn't have the xor operator...

lukamac replied:
Removed here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/772fd951234d8f0d69454d8aaec726f018f8e3a6

Scheremo commented on test/app/Makefile:
```diff
@@ -58,7 +60,10 @@ APP_SRCS += $(wildcard gen/src/*.c)
+ACCELERATOR_UPPERCASE := $(shell echo $(ACCELERATOR) | tr [:lower:] [:upper:])
+APP_CFLAGS += -DNNX_ACCELERATOR=\"$(ACCELERATOR)\" -DNNX_$(ACCELERATOR_UPPERCASE)
+
+APP_CFLAGS += -O2 -w -Wall -Werror -flto
 APP_LDFLAGS += -flto
```
I would remove `-flto` (which can cause sporadic problems on the RISCV GCC Toolchain in PULP Platform) and specify the c standard `-std=cXX` where XX is the version you prefer.

lukamac replied:
I wanted it checked with `-flto` since I believe we use it often to get better performance. So to make sure it works, at least at kernel level, I added it. @Scheremo does it make sense?
Regarding the `-std=cXX` flag, I've never used that. Do you have any recommendations?

Scheremo replied:
I think it's not unreasonable, however, flto is a program-level option, which means that the code calling into your library would have to be compiled with flto. I suppose for a test app this isn't an issue, but it might lead to people copying this Makefile into a context where flto causes issues.

Regarding the C std, I think C99 and C11 are reasonable targets, but if you really want you can also go newer. I'd suggest NOT targeting GNU extended versions (as gcc does by default!) since those can lead to tricky compliance issues with LLVM.

lukamac replied:
I removed the `-flto` here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/8232527ca95633abcee3c59f4eb648bbac257a06
I added `-std=c11` here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/df3f5ddc9d1f503e7579194a696fd731a64bf1b6 but had to check if it was compiled for GAP_SDK because they use gcc's `asm` instead of `__asm__`.
I'm not the happiest with how I check if we are compiling for GAP_SDK, so if anyone has suggestions, feel free to give them.

lukamac replied:
Removed `-std=c11` in the end and will add it for library only when I add cmake support.

Scheremo commented:
Ciao Luka,
Great work, the PR is in very good shape, I would not label any of my comments as blocking.
I think it's a good idea to resolve the TODO in `neureka/hal/neureka_task.c`, but, feel free to resolve my other comments.

It would be good to add a version branch / commit hash for the PULP-SDK / GVSoC versions this was tested with, if possible, maybe in the README? On that note, we know about a bug in the pointwise mode in Siracusa, where only unsigned weights compute correctly. Is this captured in the GVSoC version you are using?

lukamac committed:
Rename <acc>.py to <acc>MemoryLayout.py

lukamac committed:
Extract functional model from test gen

lukamac committed:
Remove conf from NeuralEngineFunctionalModel

lukamac committed:
Add isort

lukamac commented:
> Ciao Luka,
> Great work, the PR is in very good shape, I would not label any of my comments as blocking.
I think it's a good idea to resolve the TODO in neureka/hal/neureka_task.c, but, feel free to resolve my other comments.

Will do. Thanks for the comments!

> It would be good to add a version branch / commit hash for the PULP-SDK / GVSoC versions this was tested with, if possible, maybe in the README?

Yes, good idea.

> On that note, we know about a bug in the pointwise mode in Siracusa, where only unsigned weights compute correctly. Is this captured in the GVSoC version you are using?

I think it is not because I didn't do any tests with unsigned weights. @Scheremo can you point me to a gvsoc/pulp-sdk version that exhibits that behaviour?

lukamac committed:
Remove neureka siracusa clock gating

lukamac committed:
Remove inline from hal

lukamac commented on neureka/bsp/neureka_siracusa_bsp.c:
```diff
@@ -0,0 +1,95 @@
+      ~NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_CG_EN;
+}
+
+// TODO: Check if needed for neureka
+void neureka_siracusa_hci_setpriority_neureka() {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR |=
+      NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_HCI_PRIO;
+}
+
+// TODO: Check if needed for neureka
+void neureka_siracusa_hci_setpriority_core() {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR &=
+      ~NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_HCI_PRIO;
+}
+
+// TODO: Check if needed for neureka
+void neureka_siracusa_hci_reset_max_stall() {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR &=
+      ~NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_HCI_MAXSTALL;
+}
+
+// TODO: Check if needed for neureka
+void neureka_siracusa_hci_set_max_stall(uint32_t max_stall) {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR |=
+      max_stall & NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_HCI_MAXSTALL;
+}
```
@arpansur @Scheremo Is this a thing in siracusa/does it do anything?

Scheremo replied:
I am semi-sure this does do something, at least in the GVSoC port. As for Siracusa, I am not sure, I've not tried playing with it.

FrancescoConti replied:
In Siracusa: it does, although we probably do not touch this very often. It controls the QoS settings of the interconnect (see https://ieeexplore.ieee.org/document/10247945 and the figure below in particular).
On GVSOC: I am not sure it does something (I don't think it models the interconnect to a sufficient degree of accuracy to make a difference).
<img width="696" alt="image" src="https://github.com/pulp-platform/pulp-nnx/assets/7882682/83353968-8879-425f-ba29-85e305316235">

lukamac replied:
Removed todos https://github.com/pulp-platform/pulp-nnx/pull/2/commits/8b37485b2596ca44ed40436a1392b35f235a487d

lukamac committed:
Removed xor, python has xor

lukamac committed:
WIP: Add without norm_quant

lukamac committed:
Remove -flto

lukamac committed:
Add -std=c11

lukamac committed:
Move stride shift to stride2x2 function

lukamac committed:
Fix flag clear before setting

lukamac committed:
Fix normMode hardcoded to 32bit

lukamac committed:
Set quantMode in *_set_bits function

lukamac committed:
Fix output d0 stride and rename defs

lukamac committed:
Fixes to strides and stride2x2

Fixed stride2x2 mode for 32bit output.
Changed stride meaning from number of elements in that dimension, to
number of bytes between elements in that dimension. This also required a
change of dimension names.

lukamac committed:
Add no norm_quant to neureka and all the fixes too

lukamac committed:
Fix stride2x2 validity check for out channel to check stride evenness

lukamac committed:
Fix formatting

lukamac committed:
Remove TODO's cause neureka clearly needs these functions

lukamac committed:
Remove equal 0 check because that can never be the case due to the remainder function

lukamac committed:
Remove TODOs, checked padding

lukamac committed:
Rename divnceil and remainder, and add nnx_ prefix

lukamac committed:
Add citation

lukamac committed:
Add sdk and compiler commit hashes

lukamac committed:
Change task size to a define

lukamac committed:
Update supported acc supported features

lukamac committed:
Update changelog

lukamac committed:
Update pulp-sdk commit hash

lukamac committed:
Remove -std=c11 flag

lukamac committed:
Fix readme collapsable verbatim

lukamac committed:
Remove __PLATFORM__ check from the library since it's pulp-sdk specific

lukamac committed:
Change channel and bits with w_in_stride for set_ptrs
